### ISSUE TITLE
Missing declarations of genvar in unnamed generate-for loops

### ISSUE THREAD

**Issue by mglb** [chipsalliance/Surelog#3789]
Test case: https://github.com/YosysHQ/yosys/blob/f37ce5c839c5a5e21fb74046375e5928ad91d22e/tests/simple/constpower.v

command: `surelog -elabuhdm -parse -d uhdm yosys/tests/simple/constpower.v`

UHDM contains references to `i` where they are expected, but there are no declarations.
The `j` used in the inner named loop has both references and parameter declarations.
Everything works as expected when the outer loop is changed to have a name.

<details><summary>Fragment of an UHDM dump (first vpiGenScopeArray in uhdmtopModules)</summary>

```
  |vpiGenScopeArray:
  \_gen_scope_array: (work@constpower.V[0]), line:9:30, endln:12:5
    |vpiParent:
    \_module_inst: work@constpower (work@constpower), file:/constpower.v, line:1:1, endln:15:10
    |vpiName:V[0]
    |vpiFullName:work@constpower.V[0]
    |vpiGenScope:
    \_gen_scope: (work@constpower.V[0]), line:9:30, endln:12:5
      |vpiParent:
      \_gen_scope_array: (work@constpower.V[0]), line:9:30, endln:12:5
      |vpiFullName:work@constpower.V[0]
      |vpiParameter:
      \_parameter: (work@constpower.V[0].j), line:9:0
        |vpiParent:
        \_gen_scope: (work@constpower.V[0]), line:9:30, endln:12:5
        |UINT:0
        |vpiTypespec:
        \_int_typespec:
        |vpiLocalParam:1
        |vpiName:j
        |vpiFullName:work@constpower.V[0].j
      |vpiContAssign:
      \_cont_assign: , line:10:10, endln:10:72
        |vpiParent:
        \_gen_scope: (work@constpower.V[0]), line:9:30, endln:12:5
        |vpiRhs:
        \_constant:
          |vpiParent:
          \_cont_assign: , line:10:10, endln:10:72
          |vpiDecompile:0
          |vpiSize:64
          |INT:0
          |vpiConstType:7
        |vpiLhs:
        \_part_select: ys (work@constpower.ys), line:10:10, endln:10:41
          |vpiParent:
          \_cont_assign: , line:10:10, endln:10:72
          |vpiName:ys
          |vpiFullName:work@constpower.ys
          |vpiDefName:ys
          |vpiActual:
          \_logic_net: (work@constpower.ys), line:1:19, endln:1:21
          |vpiConstantSelect:1
          |vpiLeftRange:
          \_operation: , line:10:13, endln:10:27
            |vpiParent:
            \_part_select: ys (work@constpower.ys), line:10:10, endln:10:41
            |vpiOpType:24
            |vpiOperand:
            \_operation: , line:10:13, endln:10:23
              |vpiParent:
              \_operation: , line:10:13, endln:10:27
              |vpiOpType:24
              |vpiOperand:
              \_operation: , line:10:13, endln:10:16
                |vpiParent:
                \_operation: , line:10:13, endln:10:23
                |vpiOpType:25
                |vpiOperand:
                \_ref_obj: (work@constpower.V[0].ys.i), line:10:13, endln:10:14
                  |vpiParent:
                  \_operation: , line:10:13, endln:10:16
                  |vpiName:i
                  |vpiFullName:work@constpower.V[0].ys.i
                |vpiOperand:
                \_constant: , line:10:15, endln:10:16
                  |vpiParent:
                  \_operation: , line:10:13, endln:10:16
                  |vpiDecompile:8
                  |vpiSize:64
                  |UINT:8
                  |vpiConstType:9
              |vpiOperand:
              \_operation: , line:10:19, endln:10:23
                |vpiParent:
                \_operation: , line:10:13, endln:10:23
                |vpiOpType:25
                |vpiOperand:
                \_ref_obj: (work@constpower.V[0].ys.j), line:10:19, endln:10:20
                  |vpiParent:
                  \_operation: , line:10:19, endln:10:23
                  |vpiName:j
                  |vpiFullName:work@constpower.V[0].ys.j
                  |vpiActual:
                  \_parameter: (work@constpower.V[0].j), line:9:0
                |vpiOperand:
                \_constant: , line:10:21, endln:10:23
                  |vpiParent:
                  \_operation: , line:10:19, endln:10:23
                  |vpiDecompile:64
                  |vpiSize:64
                  |UINT:64
                  |vpiConstType:9
            |vpiOperand:
            \_constant: , line:10:26, endln:10:27
              |vpiParent:
              \_operation: , line:10:13, endln:10:27
              |vpiDecompile:7
              |vpiSize:64
              |UINT:7
              |vpiConstType:9
          |vpiRightRange:
          \_operation: , line:10:30, endln:10:40
            |vpiParent:
            \_part_select: ys (work@constpower.ys), line:10:10, endln:10:41
            |vpiOpType:24
            |vpiOperand:
            \_operation: , line:10:30, endln:10:33
              |vpiParent:
              \_operation: , line:10:30, endln:10:40
              |vpiOpType:25
              |vpiOperand:
              \_ref_obj: (work@constpower.V[0].ys.i), line:10:30, endln:10:31
                |vpiParent:
                \_operation: , line:10:30, endln:10:33
                |vpiName:i
                |vpiFullName:work@constpower.V[0].ys.i
              |vpiOperand:
              \_constant: , line:10:32, endln:10:33
                |vpiParent:
                \_operation: , line:10:30, endln:10:33
                |vpiDecompile:8
                |vpiSize:64
                |UINT:8
                |vpiConstType:9
            |vpiOperand:
            \_operation: , line:10:36, endln:10:40
              |vpiParent:
              \_operation: , line:10:30, endln:10:40
              |vpiOpType:25
              |vpiOperand:
              \_ref_obj: (work@constpower.V[0].ys.j), line:10:36, endln:10:37
                |vpiParent:
                \_operation: , line:10:36, endln:10:40
                |vpiName:j
                |vpiFullName:work@constpower.V[0].ys.j
                |vpiActual:
                \_parameter: (work@constpower.V[0].j), line:9:0
              |vpiOperand:
              \_constant: , line:10:38, endln:10:40
                |vpiParent:
                \_operation: , line:10:36, endln:10:40
                |vpiDecompile:64
                |vpiSize:64
                |UINT:64
                |vpiConstType:9
      |vpiContAssign:
      \_cont_assign: , line:11:10, endln:11:72
        |vpiParent:
        \_gen_scope: (work@constpower.V[0]), line:9:30, endln:12:5
        |vpiRhs:
        \_constant:
          |vpiParent:
          \_cont_assign: , line:11:10, endln:11:72
          |vpiDecompile:1
          |vpiSize:64
          |INT:1
          |vpiConstType:7
        |vpiLhs:
        \_part_select: yu (work@constpower.yu), line:11:10, endln:11:41
          |vpiParent:
          \_cont_assign: , line:11:10, endln:11:72
          |vpiName:yu
          |vpiFullName:work@constpower.yu
          |vpiDefName:yu
          |vpiActual:
          \_logic_net: (work@constpower.yu), line:1:23, endln:1:25
          |vpiConstantSelect:1
          |vpiLeftRange:
          \_operation: , line:11:13, endln:11:27
            |vpiParent:
            \_part_select: yu (work@constpower.yu), line:11:10, endln:11:41
            |vpiOpType:24
            |vpiOperand:
            \_operation: , line:11:13, endln:11:23
              |vpiParent:
              \_operation: , line:11:13, endln:11:27
              |vpiOpType:24
              |vpiOperand:
              \_operation: , line:11:13, endln:11:16
                |vpiParent:
                \_operation: , line:11:13, endln:11:23
                |vpiOpType:25
                |vpiOperand:
                \_ref_obj: (work@constpower.V[0].yu.i), line:11:13, endln:11:14
                  |vpiParent:
                  \_operation: , line:11:13, endln:11:16
                  |vpiName:i
                  |vpiFullName:work@constpower.V[0].yu.i
                |vpiOperand:
                \_constant: , line:11:15, endln:11:16
                  |vpiParent:
                  \_operation: , line:11:13, endln:11:16
                  |vpiDecompile:8
                  |vpiSize:64
                  |UINT:8
                  |vpiConstType:9
              |vpiOperand:
              \_operation: , line:11:19, endln:11:23
                |vpiParent:
                \_operation: , line:11:13, endln:11:23
                |vpiOpType:25
                |vpiOperand:
                \_ref_obj: (work@constpower.V[0].yu.j), line:11:19, endln:11:20
                  |vpiParent:
                  \_operation: , line:11:19, endln:11:23
                  |vpiName:j
                  |vpiFullName:work@constpower.V[0].yu.j
                  |vpiActual:
                  \_parameter: (work@constpower.V[0].j), line:9:0
                |vpiOperand:
                \_constant: , line:11:21, endln:11:23
                  |vpiParent:
                  \_operation: , line:11:19, endln:11:23
                  |vpiDecompile:64
                  |vpiSize:64
                  |UINT:64
                  |vpiConstType:9
            |vpiOperand:
            \_constant: , line:11:26, endln:11:27
              |vpiParent:
              \_operation: , line:11:13, endln:11:27
              |vpiDecompile:7
              |vpiSize:64
              |UINT:7
              |vpiConstType:9
          |vpiRightRange:
          \_operation: , line:11:30, endln:11:40
            |vpiParent:
            \_part_select: yu (work@constpower.yu), line:11:10, endln:11:41
            |vpiOpType:24
            |vpiOperand:
            \_operation: , line:11:30, endln:11:33
              |vpiParent:
              \_operation: , line:11:30, endln:11:40
              |vpiOpType:25
              |vpiOperand:
              \_ref_obj: (work@constpower.V[0].yu.i), line:11:30, endln:11:31
                |vpiParent:
                \_operation: , line:11:30, endln:11:33
                |vpiName:i
                |vpiFullName:work@constpower.V[0].yu.i
              |vpiOperand:
              \_constant: , line:11:32, endln:11:33
                |vpiParent:
                \_operation: , line:11:30, endln:11:33
                |vpiDecompile:8
                |vpiSize:64
                |UINT:8
                |vpiConstType:9
            |vpiOperand:
            \_operation: , line:11:36, endln:11:40
              |vpiParent:
              \_operation: , line:11:30, endln:11:40
              |vpiOpType:25
              |vpiOperand:
              \_ref_obj: (work@constpower.V[0].yu.j), line:11:36, endln:11:37
                |vpiParent:
                \_operation: , line:11:36, endln:11:40
                |vpiName:j
                |vpiFullName:work@constpower.V[0].yu.j
                |vpiActual:
                \_parameter: (work@constpower.V[0].j), line:9:0
              |vpiOperand:
              \_constant: , line:11:38, endln:11:40
                |vpiParent:
                \_operation: , line:11:36, endln:11:40
                |vpiDecompile:64
                |vpiSize:64
                |UINT:64
                |vpiConstType:9
```

</details>

**Comment by alaindargelas** [comment]
Note that Surelog never reported the parameter for i, so this is not a regression.

Also note that if you change the testcase to have explicit blocks, both loop indexes get reported as parameters:

generate
	for (i = 0; i < 2; i = i+1) begin:V
	for (j = 0; j < 2; j = j+1) begin:W
		assign ys= 1;
		
	end
end
endgenerate

**Comment by alaindargelas** [comment]
Fixed by #3791
